src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -1026,7 +1027,9 @@ module Crystal
       block_visitor.path_lookup = path_lookup || current_type
       block_visitor.block_nest = @block_nest
 
+      @while_ensure_stack.push :block
```

Comment:
`with_block` utility method looks good, however it needs to use `Array` at least in this implementation. See [line 1624](https://github.com/MakeNowJust/crystal/blob/0af105876cfb980d9984ce98c6bd3adb75d942fe/src/compiler/crystal/semantic/main_visitor.cr#L1624) for example. Not only stack top but also all stack items are important.